### ISSUE TITLE
Crashes when contiously running and switching between two fluids (LabVIEW Wrapper)

### ISSUE THREAD

**Issue by DBTaylor** [usnistgov/REFPROP-wrappers#521]
I had a system that continuously calls refprop2dll to evaluate properties of two different fluids (total 10 calls per second). This system was experiencing relatively frequent crashes (2-3 times a day). The crashes were all very different. Sometimes the app that calls refprop would crash on memory access violation. Sometimes apps that communicate with it via IPC would crash for various reasons. Removing the calls for the second fluid fixed the issue. The properties for the second fluid are not necessary for the system, so we have a solution for now. I am planning on loading a separate copy of refprop.dll for each additional fluid.

I have not been able to create a minimal reproduction. I'm not even sure there is a problem in refprop, but I figured I'd report this in case there is.

**Comment by ianhbell** [comment]
I take it this is in Labview, that makes my other proposed solution not possible. Loading multiple copies is not trivial. You need a separate DLL with a different name, as well as separate copies of the fluid files.

So I would instead propose that you make a two-component mixture, and set the mole fraction of the fluid of interest to 1.0, and 0.0 for the other one. That should be reliable.

**Comment by DBTaylor** [comment]
I have working code for multiple DLL copies (it was a pain), but I want to make sure I understand this instruction:

>as well as separate copies of the fluid files.

What makes this necessary?

**Comment by ianhbell** [comment]
Because the file reads in REFPROP are blocking, and you could get collisions if multiple copies of REFPROP are reading from the same fluid files.

It is indeed a pain.

**Comment by DBTaylor** [comment]
> Because the file reads in REFPROP are blocking, and you could get collisions if multiple copies of REFPROP are reading from the same fluid files

If I call into the multiple copies of REFPROP serially, this shouldn't be an issue then, correct?

**Comment by ianhbell** [comment]
Correct. Only parallel reads are a problem

**Comment by ianhbell** [comment]
Can you say what fluids you were switching between? Would like to try to track down this issue.

**Comment by gilberJP** [comment]
HI @ianhbell I ran into this same issue as well - we need to monitor properties for two fluids so the initial design make the calls serially - calling SETUPdll then the functions for one fluid then SETUPdll and the functions for the other fluid. I am taking the same approach as OP but before I found this post, making a separate copy of the DLL, but I was trying to debug further and found the reason the crashes were happening is handle leaks. Each time SETUPdll is called, handles are opened but never closed. Since we had two fluids, we had to call SETUPdll between each set of functions which caused our open handles to keep increasing - they didn't get closed between each call.

I decided to compare handle leaks just running the REFPROP.exe and found that handles still leak there whenever a new fluid is chosen. It stands to reason this is not a problem with LabVIEW itself but with handle leaks in the source code.

Power shell script I used to monitor the handle leaks: 

$processName = "REFPROP" # for testing refprop.exe
New-Item -ItemType Directory -Path "C:\temp" -Force | Out-Null
while ($true) {
$proc = Get-Process -Name $processName -ErrorAction SilentlyContinue
if ($proc) {
$timestamp = Get-Date -Format "yyyy-MM-dd HH:mm:ss"
$line = "$timestamp, Handles: $($proc.HandleCount), Threads: $($proc.Threads.Count), WorkingSet_MB: $([math]::Round($proc.WorkingSet64/1MB,1))"
Write-Host $line
$line | Out-File -FilePath "C:\temp\handle_monitor.csv" -Append
} else {
Write-Host "Process '$processName' not found"
}
Start-Sleep -Seconds 30
}
